crypto/src/hash/rescue_prime/mod.rs

```diff
@@ -0,0 +1,816 @@
+                *new_value += self.mds_matrix[i][j] * state_value;
+            }
+        }
+
+        new_state
+    }
+    fn mds_ntt(&self, state: &[Fp]) -> Vec<Fp> {
+        let m = state.len();
+        let omega = match m {
+            12 => Fp::from(281474976645120u64),
+            16 => Fp::from(17293822564807737345u64),
+            _ => panic!("Unsupported state size for NTT"),
+        };
+
+        let mds_ntt = ntt(&self.mds_vector, omega);
+        let state_rev: Vec<Fp> = iter::once(state[0])
+            .chain(state[1..].iter().rev().cloned())
+            .collect();
+        let state_ntt = ntt(&state_rev, omega);
+
+        let mut product_ntt = vec![Fp::zero(); m];
+        for i in 0..m {
+            product_ntt[i] = mds_ntt[i] * state_ntt[i];
+        }
+
+        let omega_inv = omega.inv().unwrap();
+        let result = intt(&product_ntt, omega_inv);
+
+        iter::once(result[0])
+            .chain(result[1..].iter().rev().cloned())
+            .collect()
+    }
+
+    fn mds_karatsuba(&self, state: &[Fp]) -> Vec<Fp> {
+        let m = state.len();
+        let mds_rev: Vec<Fp> = iter::once(self.mds_vector[0])
+            .chain(self.mds_vector[1..].iter().rev().cloned())
+            .collect();
+
+        let conv = karatsuba(&mds_rev, state);
+
+        let mut result = vec![Fp::zero(); m];
+        result[..m].copy_from_slice(&conv[..m]);
+        for i in m..conv.len() {
+            result[i - m] += conv[i];
+        }
+
+        result
+    }
+
+    fn apply_mds(&self, state: &mut [Fp]) {
+        let new_state = match self.mds_method {
+            MdsMethod::MatrixMultiplication => self.mds_matrix_vector_multiplication(state),
+            MdsMethod::Ntt => self.mds_ntt(state),
+            MdsMethod::Karatsuba => self.mds_karatsuba(state),
+        };
+        state.copy_from_slice(&new_state);
+    }
+
+    fn add_round_constants(&self, state: &mut [Fp], round: usize) {
+        let m = self.m;
+        let round_constants = &self.round_constants;
+        for j in 0..m {
+            state[j] += round_constants[round * 2 * m + j];
+        }
+    }
+
+    fn add_round_constants_second(&self, state: &mut [Fp], round: usize) {
+        let m = self.m;
+        let round_constants = &self.round_constants;
+        for j in 0..m {
+            state[j] += round_constants[round * 2 * m + m + j]
+        }
+    }
+
+    pub fn permutation(&self, state: &mut [Fp]) {
+        let num_rounds = NUM_FULL_ROUNDS;
+        for round in 0..num_rounds {
+            self.apply_mds(state);
+            self.add_round_constants(state, round);
+            Self::apply_sbox(state);
+            self.apply_mds(state);
+            self.add_round_constants_second(state, round);
+            Self::apply_inverse_sbox(state);
+        }
+    }
+
+    pub fn hash(&self, input_sequence: &[Fp]) -> Vec<Fp> {
+        let m = self.m;
+        let capacity = self.capacity;
+        let rate = self.rate;
```

Comment:
We don't need these variable declarations here (we do not have any advantage in doing this).

Reply:
Agree, I've removed the variable declarations since they weren't necessary